mypy/checker.py

```diff
@@ -1400,36 +1470,43 @@ def visit_return_stmt(self, s: ReturnStmt) -> Type:
+                return_type = self.get_generator_return_type(self.return_types[-1])
+            else:
+                return_type = self.return_types[-1]
+
             if s.expr:
                 # Return with a value.
-                typ = self.accept(s.expr, self.return_types[-1])
+                typ = self.accept(s.expr, return_type)
                 # Returning a value of type Any is always fine.
-                if not isinstance(typ, AnyType):
-                    if isinstance(self.return_types[-1], Void):
-                        # FuncExpr (lambda) may have a Void return.
-                        # Function returning a value of type None may have a Void return.
-                        if (not isinstance(self.function_stack[-1], FuncExpr) and
-                                not isinstance(typ, NoneTyp)):
-                            self.fail(messages.NO_RETURN_VALUE_EXPECTED, s)
-                    else:
-                        if self.function_stack[-1].is_coroutine:
-                            # Something similar will be needed to mix return and yield.
-                            # If the function is a coroutine, wrap the return type in a Future.
-                            typ = self.wrap_generic_type(cast(Instance, typ),
-                                                         cast(Instance, self.return_types[-1]),
-                                                         'asyncio.futures.Future', s)
-                        self.check_subtype(
-                            typ, self.return_types[-1], s,
-                            messages.INCOMPATIBLE_RETURN_VALUE_TYPE
-                            + ": expected {}, got {}".format(self.return_types[-1], typ)
-                        )
+                if isinstance(typ, AnyType):
+                    return None
+
+                if isinstance(return_type, Void):
+                    # FuncExpr (lambda) may have a Void return.
+                    # Function returning a value of type None may have a Void return.
+                    if isinstance(self.function_stack[-1], FuncExpr) or isinstance(typ, NoneTyp):
+                        return None
+
+                    self.fail(messages.NO_RETURN_VALUE_EXPECTED, s)
+                else:
+                    self.check_subtype(
+                        typ, return_type, s,
+                        messages.INCOMPATIBLE_RETURN_VALUE_TYPE
+                        + ": expected {}, got {}".format(return_type, typ)
+                    )
             else:
-                # Return without a value. It's valid in a generator and coroutine function.
-                if (not self.function_stack[-1].is_generator and
-                        not self.function_stack[-1].is_coroutine):
-                    if (not isinstance(self.return_types[-1], Void) and
-                            self.typing_mode_full()):
-                        self.fail(messages.RETURN_VALUE_EXPECTED, s)
+                # return without a value
```

Comment:
I prefer comments (esp. whole-line comments) to always be properly capitalized sentences ending with correct punctuation.